Restic on APFS needs small adjustments

## Output of `restic version`

```
$  ./restic version
debug enabled
restic 0.6.1 (v0.6.1-73-gaabe2a0-dirty)
compiled with go1.8.1 on darwin/amd64
```

## How did you start restic exactly? (Include the complete command line)

```
$ gb test cmds/restic
```

## What backend/server/service did you use?

Local FS.

## Expected behavior

Tests should pass.

## Actual behavior

Using 10.13 beta on APFS, there is a precision discreptancy. I don't know what the deal is with the lchown.

```
$ gb test cmds/restic
# cmds/restic
Now serving the repository at /var/folders/8n/00j6g_xx08qb21g22s_jj7sh0000gn/T/restic-test-mount-242244809
Don't forget to umount after quitting!
Now serving the repository at /var/folders/8n/00j6g_xx08qb21g22s_jj7sh0000gn/T/restic-test-mount-242244809
Don't forget to umount after quitting!
Now serving the repository at /var/folders/8n/00j6g_xx08qb21g22s_jj7sh0000gn/T/restic-test-mount-242244809
Don't forget to umount after quitting!
Now serving the repository at /var/folders/8n/00j6g_xx08qb21g22s_jj7sh0000gn/T/restic-test-mount-242244809
Don't forget to umount after quitting!
Now serving the repository at /var/folders/8n/00j6g_xx08qb21g22s_jj7sh0000gn/T/restic-test-mount-055822315
Don't forget to umount after quitting!
error for testdata/0/0/9/37: Lstat: lstat /var/folders/8n/00j6g_xx08qb21g22s_jj7sh0000gn/T/restic-test-149018899/testdata/0/0/9/37: no such file or directory
error walking dir testdata/0/0/9/37: Lstat: lstat /var/folders/8n/00j6g_xx08qb21g22s_jj7sh0000gn/T/restic-test-049004756/testdata/0/0/9/37: no such file or directory
error walking dir 0/9: Open: open /var/folders/8n/00j6g_xx08qb21g22s_jj7sh0000gn/T/restic-test-184104134/testdata/0/0/9: no such file or directory
 ID          User        Host        Created
----------------------------------------------------------------------
*dc3fff31    lloeki      Galileo.local  2017-07-06 17:28:03
foo: ModTime does not match (2017-07-06 17:28:04.513803084 +0200 CEST != 2017-07-06 17:28:04.513804 +0200 CEST)
%foo
foo/bar: ModTime does not match (2017-07-06 17:28:08.71747536 +0200 CEST != 2017-07-06 17:28:08.717476 +0200 CEST)
%foo/bar
foo/bar/testfile0: ModTime does not match (2017-07-06 17:28:05.227860691 +0200 CEST != 2017-07-06 17:28:05.227861 +0200 CEST)
%foo/bar/testfile0
foo/bar/testfile1: ModTime does not match (2017-07-06 17:28:05.71512518 +0200 CEST != 2017-07-06 17:28:05.715126 +0200 CEST)
%foo/bar/testfile1
foo/bar/testfile2: ModTime does not match (2017-07-06 17:28:05.807257682 +0200 CEST != 2017-07-06 17:28:05.807258 +0200 CEST)
%foo/bar/testfile2
foo/bar/testfile3: ModTime does not match (2017-07-06 17:28:06.307714912 +0200 CEST != 2017-07-06 17:28:06.307715 +0200 CEST)
%foo/bar/testfile3
foo/bar/testfile4: ModTime does not match (2017-07-06 17:28:07.078611832 +0200 CEST != 2017-07-06 17:28:07.078612 +0200 CEST)
%foo/bar/testfile4
foo/bar/testfile5: ModTime does not match (2017-07-06 17:28:07.604224415 +0200 CEST != 2017-07-06 17:28:07.604225 +0200 CEST)
%foo/bar/testfile5
foo/bar/testfile6: ModTime does not match (2017-07-06 17:28:07.971472573 +0200 CEST != 2017-07-06 17:28:07.971473 +0200 CEST)
%foo/bar/testfile6
foo/bar/testfile7: ModTime does not match (2017-07-06 17:28:08.023946232 +0200 CEST != 2017-07-06 17:28:08.023947 +0200 CEST)
%foo/bar/testfile7
foo/bar/testfile8: ModTime does not match (2017-07-06 17:28:08.708402862 +0200 CEST != 2017-07-06 17:28:08.708403 +0200 CEST)
%foo/bar/testfile8
foo/bar/testfile9: ModTime does not match (2017-07-06 17:28:09.223617926 +0200 CEST != 2017-07-06 17:28:09.223618 +0200 CEST)
%foo/bar/testfile9
integration_test.go:893: directories are not equal

--- FAIL: TestRestore (6.17s)
	testing.go:29: using low-security KDF parameters for test
	testing.go:29: using low-security KDF parameters for test
	lock.go:77: setting lock timeout to 0s
	integration_test.go:51: repository initialized at /var/folders/8n/00j6g_xx08qb21g22s_jj7sh0000gn/T/restic-test-030682587/repo
	integration_test.go:55: backing up [/var/folders/8n/00j6g_xx08qb21g22s_jj7sh0000gn/T/restic-test-030682587/testdata]
index d5697fe0 has old format
There were 4 errors
index d5697fe0 has old format
ignoring error for /var/folders/8n/00j6g_xx08qb21g22s_jj7sh0000gn/T/restic-test-265310113/restore/testdir: Lchown: lchown /var/folders/8n/00j6g_xx08qb21g22s_jj7sh0000gn/T/restic-test-265310113/restore/testdir: operation not permitted
ignoring error for /var/folders/8n/00j6g_xx08qb21g22s_jj7sh0000gn/T/restic-test-265310113/restore/testdir/bar: Lchown: lchown /var/folders/8n/00j6g_xx08qb21g22s_jj7sh0000gn/T/restic-test-265310113/restore/testdir/bar: operation not permitted
ignoring error for /var/folders/8n/00j6g_xx08qb21g22s_jj7sh0000gn/T/restic-test-265310113/restore/testdir/bar2: Lchown: lchown /var/folders/8n/00j6g_xx08qb21g22s_jj7sh0000gn/T/restic-test-265310113/restore/testdir/bar2: operation not permitted
ignoring error for /var/folders/8n/00j6g_xx08qb21g22s_jj7sh0000gn/T/restic-test-265310113/restore/testdir/baz: Lchown: lchown /var/folders/8n/00j6g_xx08qb21g22s_jj7sh0000gn/T/restic-test-265310113/restore/testdir/baz: operation not permitted
There were 4 errors
ignoring error for /var/folders/8n/00j6g_xx08qb21g22s_jj7sh0000gn/T/restic-test-062584457/restore/random: Lchown: lchown /var/folders/8n/00j6g_xx08qb21g22s_jj7sh0000gn/T/restic-test-062584457/restore/random: operation not permitted
There were 1 errors
ignoring error for /var/folders/8n/00j6g_xx08qb21g22s_jj7sh0000gn/T/restic-test-062584457/restore/random: Lchown: lchown /var/folders/8n/00j6g_xx08qb21g22s_jj7sh0000gn/T/restic-test-062584457/restore/random: operation not permitted
There were 1 errors
ignoring error for /var/folders/8n/00j6g_xx08qb21g22s_jj7sh0000gn/T/restic-test-062584457/restore/random: Lchown: lchown /var/folders/8n/00j6g_xx08qb21g22s_jj7sh0000gn/T/restic-test-062584457/restore/random: operation not permitted
There were 1 errors
ignoring error for /var/folders/8n/00j6g_xx08qb21g22s_jj7sh0000gn/T/restic-test-062584457/restore/random: Lchown: lchown /var/folders/8n/00j6g_xx08qb21g22s_jj7sh0000gn/T/restic-test-062584457/restore/random: operation not permitted
There were 1 errors
FAIL
FATAL: command "test" failed: [/var/folders/8n/00j6g_xx08qb21g22s_jj7sh0000gn/T/gb442135313/cmds/restic/testmain/_test/testmain]: exit status 1
```

## Steps to reproduce the behavior

run tests on 10.13 with APFS

## Do you have any idea what may have caused this?

APFS is known to have increased timestamp precision down to nanosecond.

@fd0 sure, output:

```
> DEBUG_LOG=/tmp/log gb test -tags debug -v -run 'TestRestore$' cmds/restic
restic/repository
restic/checker
github.com/pkg/sftp
restic/backend/sftp
restic/backend/location
# cmds/restic
debug log file /tmp/log
debug enabled
=== RUN   TestRestore
foo: ModTime does not match (2017-07-08 12:31:40.164913568 +0200 CEST != 2017-07-08 12:31:40.164914 +0200 CEST)
%foo
foo/bar: ModTime does not match (2017-07-08 12:31:44.814164684 +0200 CEST != 2017-07-08 12:31:44.814165 +0200 CEST)
%foo/bar
foo/bar/testfile0: ModTime does not match (2017-07-08 12:31:40.913921105 +0200 CEST != 2017-07-08 12:31:40.913922 +0200 CEST)
%foo/bar/testfile0
foo/bar/testfile1: ModTime does not match (2017-07-08 12:31:41.471042803 +0200 CEST != 2017-07-08 12:31:41.471043 +0200 CEST)
%foo/bar/testfile1
foo/bar/testfile2: ModTime does not match (2017-07-08 12:31:41.574751352 +0200 CEST != 2017-07-08 12:31:41.574752 +0200 CEST)
%foo/bar/testfile2
foo/bar/testfile3: ModTime does not match (2017-07-08 12:31:42.152517535 +0200 CEST != 2017-07-08 12:31:42.152518 +0200 CEST)
%foo/bar/testfile3
foo/bar/testfile4: ModTime does not match (2017-07-08 12:31:43.014842612 +0200 CEST != 2017-07-08 12:31:43.014843 +0200 CEST)
%foo/bar/testfile4
foo/bar/testfile5: ModTime does not match (2017-07-08 12:31:43.602873562 +0200 CEST != 2017-07-08 12:31:43.602874 +0200 CEST)
%foo/bar/testfile5
foo/bar/testfile6: ModTime does not match (2017-07-08 12:31:44.001077844 +0200 CEST != 2017-07-08 12:31:44.001078 +0200 CEST)
%foo/bar/testfile6
foo/bar/testfile7: ModTime does not match (2017-07-08 12:31:44.058968731 +0200 CEST != 2017-07-08 12:31:44.058969 +0200 CEST)
%foo/bar/testfile7
foo/bar/testfile8: ModTime does not match (2017-07-08 12:31:44.798702627 +0200 CEST != 2017-07-08 12:31:44.798703 +0200 CEST)
%foo/bar/testfile8
foo/bar/testfile9: ModTime does not match (2017-07-08 12:31:45.355062584 +0200 CEST != 2017-07-08 12:31:45.355063 +0200 CEST)
%foo/bar/testfile9
integration_test.go:893: directories are not equal

--- FAIL: TestRestore (7.62s)
	testing.go:29: using low-security KDF parameters for test
	testing.go:29: using low-security KDF parameters for test
	lock.go:77: setting lock timeout to 0s
	integration_test.go:51: repository initialized at /var/folders/rm/7k1v_5pj6sxgm2mq0k9fpj6h0000gn/T/restic-test-184666079/repo
	integration_test.go:55: backing up [/var/folders/rm/7k1v_5pj6sxgm2mq0k9fpj6h0000gn/T/restic-test-184666079/testdata]
FAIL
FATAL: command "test" failed: [/var/folders/rm/7k1v_5pj6sxgm2mq0k9fpj6h0000gn/T/gb422824338/cmds/restic/testmain/_test/testmain-debug -test.v=true -test.run=TestRestore$]: exit status 1
```

Contents of `/tmp/log` [attached](https://github.com/restic/restic/files/1132899/log.txt)

--- a/cmd/restic/integration_helpers_test.go
+++ b/cmd/restic/integration_helpers_test.go
@@ -71,7 +71,17 @@ func sameModTime(fi1, fi2 os.FileInfo) bool {
 		}
 	}
 
-	return fi1.ModTime().Equal(fi2.ModTime())
+	same := fi1.ModTime().Equal(fi2.ModTime())
+	if !same && runtime.GOOS == "darwin" {
+		// Allow up to 1μs difference, because macOS <10.13 cannot restore
+		// with nanosecond precision and the current version of Go (1.9.2)
+		// does not yet support the new syscall. (#1087)
+		mt1 := fi1.ModTime()
+		mt2 := fi2.ModTime()
+		usecDiff := (mt1.Nanosecond()-mt2.Nanosecond())/1000 + (mt1.Second()-mt2.Second())*1000000
+		same = usecDiff <= 1 && usecDiff >= -1
+	}
+	return same
 }
 
 // directoriesEqualContents checks if both directories contain exactly the same

